### PR TITLE
More date autoconditions

### PR DESCRIPTION
**Pull request by warp-core**
**Feature:** This PR implements the feature request detailed and discussed in issue #6553

## Feature Details
Adds 4 new autoconditions:
`"days since year start"` - the number of days since the beginning of the year; the number of days in previous months plus the current day of the month.
`"days until year end"` - the number of days until the end of hte year; 365 or 366 (depending on if this is a leap year) minus the current value of `"days since year start"`.
`"days since epoch"` - the number of days since the "epoch" (1/1/1).
`"days since start"` - the number of days since the day this pilot started on. If the current date is somehow before the date the pilot started on, this value will be negative.

Also adds corresponding methods to Date.

### CODE REVIEW

**Comment by Amazinite** [comment]
It was also mentioned in #6553 the possibility of making an "epoch" condition, which already has a function in Date.

**Comment by warp-core** [comment]
Just `"epoch"` or `"days since epoch"` or...?

**Comment by Zitchas** [comment]
I gather, from the description, that the game actually calculates whether or not a particular year is a leap year?
What about having the starting date as a condition? (I know it's recorded as part of the starting data, but I don't think we can check that, can we?

**Comment by warp-core** [comment]
> I gather, from the description, that the game actually calculates whether or not a particular year is a leap year? What about having the starting date as a condition? (I know it's recorded as part of the starting data, but I don't think we can check that, can we?

Yes, but probably in another PR.
You could also just set conditions with the day, month, and year of the start in the start.
With this PR, it'll also be possible to set a conditions with the epoch value at the start to make, for example, having something occur x days after the start much easier.

**Comment by warp-core** [comment]
> So... What is the 'epoch' ? I gather it is probably the start of the game, is there any possibility it might be something else, or changed? Any possibility that new epochs start or something?

Days since 1/1/1.
It is currently only used internally, but with this PR, you could set a condition to the current epoch value at the start of the game and then later compare it to the current date and require a minimum difference to do something.

**Comment by Zitchas** [comment]
OK. So we start off at something a bit over 1.1 million, then? (days since 1/1/1, most starts are around 3014...)

**Comment by Amazinite** [comment]
> It is currently only used internally, but with this PR, you could set a condition to the current epoch value at the start of the game and then later compare it to the current date and require a minimum difference to do something.

I think it'd be simpler for this condition to just consider the epoch to be the start date on the save, since we can't really compare the current date and the start date for old files, unless we make the start date its own condition.

**Comment by Zitchas** [comment]
So, why can't this "epoch" value just be labeled "days since start" ? That'd be *much* clearer.

**Comment by Zireael07** [comment]
@Zitchas: I am kinda confused why we need epoch (I'm assuming 1/1/1 is January 1 CE) at all? Days since start should be both more understandable and cover everything - and if we need time to something that happened before start, well, negative integers exist :)

**Comment by warp-core** [comment]
> unless we make the start date its own condition.

I think that's what Zitchas is asking for.

> So, why can't this "epoch" value just be labeled `"days since start"` and `"start date"` respectively? That'd be much clearer.

Because that's not what it currently is.
Also, what would `"start date"` be? The epoch value of the start date? A single value isn't really a "date". Would you want separate `"start: day"`, "`start: month"` and `"start: year"` conditions as well?

**Comment by Zireael07** [comment]
Most games I know either leave the start date up to the player's imagination (e,g, "sometime in 2030" given in someone's story/background snippets) or pick a specific date (usually, on the programming side, stored in Unix time) and then count in days_since_start, for both approaches.

**Comment by Zitchas** [comment]
Alright, I think we have some miscommunication, then. It was said that the Epoch value is the start date, well, the start date is 3013-Nov-16 for our current default new pilot. But it sounds like you are saying that this is *not* what the epoch is. This leads me to guess that what is being referred to as the "Start date" isn't actually the start of the game, but rather referring to the baseline that defines when the game's calendar started counting. If we start counting at 1/1/1, then, according to the date-and-time calculators, our value for "time since arbitrary reference point."

In other words, we need some clarity as to what exactly is being measured, because so far, "epoch" means a lot of different things, and so far, I'm just making guesses about what it might be. Guesses aren't useful for actually using an attribute, so some very clear and specific definitions are required.

**Comment by warp-core** [comment]
> It was said that the Epoch value is the start date

Where? I don't recall saying that and it doesn't look to me like anyone else here has said that.

From my earlier comment:

> > So... What is the 'epoch' ? I gather it is probably the start of the game, is there any possibility it might be something else, or changed? Any possibility that new epochs start or something?

> Days since 1/1/1.

Derpy has suggested that the epoch condition should instead be `"days since start"`, not that that's what it currently is.

**Comment by Zitchas** [comment]
Ok, well, no-one has clearly stated what an epoch *is*, then. I just keep feeling like it's assumed everyone knows, or if it's some value that is already defined or something.

**Comment by warp-core** [comment]
I thought:
> `"days since epoch"` - the number of days since the "epoch" (1/1/1).
Was clear enough.
To be clearer: this is the number of days since the 1st of January in the year 1.

File: source/PlayerInfo.cpp [hunk 1]
```diff
@@ -2709,6 +2709,21 @@ void PlayerInfo::RegisterDerivedConditions()
...
+	auto &&daysUntilYearEndProvider = conditions.GetProviderNamed("days until year end");
+	daysUntilYearEndProvider.SetGetFunction([this](const string &name) { return date.DaysUntilYearEnd(); });
+
+	auto &&daysSinceEpochProvider = conditions.GetProviderNamed("days since epoch");
+	daysSinceEpochProvider.SetGetFunction([this](const string &name) { return date.DaysSinceEpoch(); });
+
+	auto &&daysSinceStartProvider = conditions.GetProviderNamed("days since start");
+	daysSinceStartProvider.SetGetFunction([this](const string &name)
+	{
+		return StartData().GetDate().DaysSinceEpoch() - date.DaysSinceEpoch();
```
Comment (warp-core) [inline source/PlayerInfo.cpp:2724]:
```suggestion
		return date.DaysSinceEpoch() - StartData().GetDate().DaysSinceEpoch();
```
